Amazon Opensearch support

AhmadMasry opened the pull request:
# What Does This Do
Add instrumentation support for Amazon Opensearch

mcculls commented on dd-java-agent/instrumentation/opensearch/rest/src/main/java/datadog/trace/instrumentation/opensearch/OpensearchRestClientInstrumentation.java:
```diff
@@ -0,0 +1,110 @@
+  @Override
+  public ElementMatcher<ClassLoader> classLoaderMatcher() {
+    // Avoid matching pre-ES7 releases which have their own instrumentations.
+    return hasClassNamed("org.opensearch.client.RestClient$InternalRequest");
```
This `classLoaderMatcher` method can be removed since we only have one opensearch instrumentation
(it was only necessary with elasticsearch to distinguish between the different elasticsearch instrumentations)

mcculls commented on dd-java-agent/instrumentation/opensearch/rest/src/main/java/datadog/trace/instrumentation/opensearch/RestResponseListener.java:
```diff
@@ -0,0 +1,43 @@
+import org.opensearch.client.Response;
+import org.opensearch.client.ResponseListener;
+
+/** This class is identical to version 6's instrumentation. */
```
This comment can be removed since we only have one opensearch instrumentation

mcculls commented on dd-java-agent/instrumentation/opensearch/src/main/java/datadog/trace/instrumentation/opensearch/OpensearchTransportClientDecorator.java:
```diff
@@ -0,0 +1,82 @@
+  public AgentSpan onRequest(final AgentSpan span, final Class action, final Class request) {
+    if (action != null) {
+      String actionName = action.getSimpleName();
+      // ES 7.9 internally changes PutMappingAction to AutoPutMappingAction for
```
s/ES 7.9/Opensearch/ ?

mcculls commented:
Hi @AhmadMasry - to complete this instrumentation you'll also need to copy the classes from `dd-java-agent/instrumentation/elasticsearch/transport-7.3/` and refactor them to use the opensearch package.

You can copy them into your new `dd-java-agent/instrumentation/opensearch/transport` module along with the tests from  `transport-7.3` - you'll then need to marge `transport-7.3/build.gradle` with your new `opensearch/build.gradle`

Once those classes have been copied over then it should be good to merge - thanks!

AhmadMasry commented:
Hi @mcculls 
Moving transport-7.3 was a bit tricky, where it is not supported at all in OpenSearch 2 and many needed functionalities are removed and deprecated in OpenSearch 1. 
I removed the incompatible code, modified muzzle and refactored tests so they can pass.
One strange thing is the need to add multiple Thread.sleep(10000) before any trace evaluation as all of them where timing out in any assertTraces or TEST_WRITER.waitForTraces.
I will be great if someone can check why they are timing out to help me remove these sleeps.
Thanks

AhmadMasry commented:
BTW, removed the sleeps, and the tests are successful on CircleCI, it seems a local issue on my laptop.

mcculls commented:
FYI this change on master should fix the elasticsearch muzzle failure: https://github.com/DataDog/dd-trace-java/pull/5282

AhmadMasry commented:
Only system tests are failing now.

AhmadMasry commented:
Hi @smola and @mcculls 
Did anyone have the chance to review my work, or it can be included in the agent or not?
thank you for your efforts
